nucypher/policy/orders.py

```diff
@@ -38,310 +38,260 @@
-                (stamp, 'ursula_stamp', PublicKey.serialized_size()),
-                (encrypted_kfrag, 'encrypted_kfrag', AuthorizedKeyFrag.ENCRYPTED_SIZE)
-                # NOTE: ursula_identity_evidence has a default value of b'' for federated mode.
-            )
+    def __init__(self, treasure_map: TreasureMap, retrieval_kits: Sequence['RetrievalKit']):
 
-            for parameter, name, expected_length in expected_lengths:
-                if len(parameter) != expected_length:
-                    raise ValueError(f"{name} must be of length {expected_length}, but it's {len(parameter)}")
+        # Record the retrieval kits order
+        self._capsules = [retrieval_kit.capsule for retrieval_kit in retrieval_kits]
 
-            task_specification = (
-                bytes(self.capsule),
-                stamp,
-                identity_evidence
-            )
-            return b''.join(task_specification)
-
-        def __bytes__(self):
-            data = bytes(self.capsule) + bytes(self.signature)
-            if self.cfrag and self.cfrag_signature:
-                data += bytes(self.cfrag) + bytes(self.cfrag_signature)
-            return data
-
-        @classmethod
-        def from_bytes(cls, data: bytes):
-            capsule, signature, remainder = cls.input_splitter(data, return_remainder=True)
-            if remainder:
-                cfrag, reencryption_signature = cls.output_splitter(remainder)
-                return cls(capsule=capsule, signature=signature, cfrag=cfrag, cfrag_signature=reencryption_signature)
-            else:
-                return cls(capsule=capsule, signature=signature)
-
-        def attach_work_result(self, cfrag, cfrag_signature) -> None:
-            self.cfrag = cfrag
-            self.cfrag_signature = cfrag_signature
+        self._threshold = treasure_map.threshold
 
-    def __init__(self,
-                 bob: 'Bob',
-                 hrac: HRAC,
-                 encrypted_kfrag: bytes,
-                 alice_verifying_key: bytes,
-                 publisher_verifying_key: bytes,
-                 tasks: dict,
-                 receipt_signature: Signature,
-                 ursula: Optional['Ursula'] = None):
+        # Records the retrieval results, indexed by capsule
+        self._results = {retrieval_kit.capsule: {}
+                         for retrieval_kit in retrieval_kits} # {capsule: {ursula_address: cfrag}}
 
-        self.bob = bob
-        self.hrac = hrac
-        self.ursula = ursula
-        self.encrypted_kfrag = encrypted_kfrag
-        self.publisher_verifying_key = publisher_verifying_key
-        self.alice_verifying_key = alice_verifying_key
-        self.tasks = tasks
-        self.receipt_signature = receipt_signature  # not a blockchain receipt
-        self.completed = False
+        # Records the addresses of Ursulas that were already queried, indexed by capsule.
+        self._queried_addresses = {retrieval_kit.capsule: retrieval_kit.queried_addresses
```

Comment:
Side effect of changing `RetrievalKit.queried_addresses` from a `Set` to `List` causes this line to have subsequent effects on:
1. L102 where we try to do `ursulas_to_contact_last |= queried_addresses` but would have to do `ursulas_to_contact_last |= set(queried_addresses)` instead
2. L136 where we try to do `self._queried_addresses[capsule].add(work_order.ursula_address)` but would have to do `self._queried_addresses[capsule].append(work_order.ursula_address)` instead.

Some of my Porter tests in #2768 are now failing because of it.
Either we use a set here:`self._queried_addresses = {retrieval_kit.capsule: set(retrieval_kit.queried_addresses)`, or we make the above changes.

Reply:
Good catch, the last is especially bed since I'm mutating the externally passed objects. Fixed.